### ISSUE TITLE
doc_count_error_upper_bound is under-estimated for the Terms Aggregation and therefore is not an upper bound

### ISSUE THREAD

**Issue by diogoholanda** [elastic/elasticsearch#35987]
This bug report is based on the documentation of the current Elasticsearch version (6.5).

**Description of the problem including expected versus actual behavior**:
I the documentation for how "doc_count_error_upper_bound" is calculated (https://www.elastic.co/guide/en/elasticsearch/reference/6.5/search-aggregations-bucket-terms-aggregation.html#_calculating_document_count_error) it is said that " This is calculated as the sum of the document count from the last term returned from each shard. For the example given above the value would be 46 (2 + 15 + 29)."

If you look at the numeric example given (https://www.elastic.co/guide/en/elasticsearch/reference/6.5/search-aggregations-bucket-terms-aggregation.html#search-aggregations-bucket-terms-aggregation-approximate-counts), you can see that the error could actually be above 46.

Take Product F, for instance. It was not returned by shard A, so its value is at most 2. It was returned by shard B with a value of 17. It was not returned by shard C, so its value is at most 29. Therefore, its largest possible doc count is 2 + 17 + 29 = 48.

It seems that, to calculate the real "doc_count_error_upper_bound", we should consider both:
- The maximum possible doc_count for terms not returned by any bucket (this is currently considered)
- The maximum possible doc_count for terms returned by some buckets but that did not make the final list (this is not currently considered)

**Comment by diogoholanda** [comment]
@cbuescher what's the best way to move this forward?

**Comment by cbuescher** [comment]
@diogoholanda By labeling this issue it should be in the queue for one of our area teams, so it should get picked up soon.

**Comment by polyfractal** [comment]
We chatted about this earlier in the week and I just didn't get a chance to respond.  Sorry!

It does seem like a bug.  Looking at the code, what we're giving the user is basically the maximum potential doc_count of a term which didn't make _any of the shards_ top `n`.  Which isn't quite what we intended.  It should be the maximum potential doc_countn of a term which didn't make the final top `n`, not just any particular shard.

To fix this, I think we need to also track the max doc_count of buckets pruned during the final reduction and use that to modify the error (if applicable).

That said, the goal of the error is to express to the user an order of "wrongness" in their results, as a way to gauge how much to trust it.  So I could see it going three ways:

1. It's a bug, would be nice to fix but not critical since the error should be reasonably close except for the most antagonist datasets
2. Fix the bug so it's accurate
3. Find some way to express the error in a unit-less quantity, so that the user gets a sense of magnitude without actual hard numbers (percentage, or something?  Not sure).

I'm going to label this as a `bug` and `help-wanted`.  If you'd like to work on a PR to fix it we'd love to work with you to get it fixed!
